src/constructs/core/migrating.ts

```diff
@@ -0,0 +1,41 @@
+import type { CfnElement, IConstruct } from "@aws-cdk/core";
+import type { GuStack } from "./stack";
+
+export interface GuMigratingResource {
+  existingLogicalId?: string;
+}
+
+export const GuMigratingResource = {
+  setLogicalId<T extends IConstruct>(
+    construct: T,
+    { migratedFromCloudFormation }: GuStack,
+    { existingLogicalId }: GuMigratingResource
+  ): void {
+    const overrideLogicalId = (logicalId: string) => {
+      const defaultChild = construct.node.defaultChild as CfnElement;
+      defaultChild.overrideLogicalId(logicalId);
+    };
+
+    const id = construct.node.id;
+
+    if (!migratedFromCloudFormation && existingLogicalId) {
```

Comment:
I hadn't really thought about it before, but I guess this could be a valid thing to do (in the future) if the user has created a resource via one cdk template and is now trying to move it into another? It's a bit of an edge case though, so feel free to ignore!

Reply:
Actually I'm not sure this makes sense at all - definitely ignore it!

Reply:
It's odd, right?! This behaviour is implicitly in place in things like [GuPolicy](https://github.com/guardian/cdk/blob/60cbf21698b3ddbb8cfd23556fd2af886e9551c7/src/constructs/iam/policies/base-policy.ts#L15-L18) as it's not checking the value of `migratedFromCloudFormation`.

It sounds like you're suggesting we only ever observe `existingLogicalId` if `migratedFromCloudFormation` is explicitly true? That makes perfect sense to me!

Reply:
Yeah, I guess we really want to account for _any_ migration, not just from a CFN YAML but also from CDK. 

I'm wondering if there's ever a situation in which you're _not_ migrating the stack, but you _are_ migrating the resource. If there isn't, then we can definitely follow the suggestion above to only observe `existingLogicalId` if `migratedFromCloudFormation` is true.

Reply:
Yeah, I guess we really want to account for _any_ migration, not just from a CFN YAML but also from CDK. 

I'm wondering if there's ever a situation in which you're _not_ migrating the stack, but you _are_ migrating the resource. If there isn't, then we can definitely follow the suggestion above to only observe `existingLogicalId` if `migratedFromCloudFormation` is true.

Reply:
> Yeah, I guess we really want to account for any migration, not just from a CFN YAML but also from CDK.

I suspect a CDK to Gu-CDK migration would generate the same ID if the output of the construct looks the same. For example, if migrating a policy and the policy documents matched.

If we do migrate between CDK and Gu-CDK, it would be intuitive to set `migratedFromCloudFormation` to true as it _is_ a migration. That said, I'm increasingly thinking we should mandate `migratedFromCloudFormation` be true for `existingLogicalId` to be observed.